session/utils_test.go

```diff
@@ -0,0 +1,373 @@
+			err:    "",
+		},
+		{
+			name:   "bool on",
+			input:  "select %?",
+			params: []interface{}{true},
+			output: "select 1",
+			err:    "",
+		},
+		{
+			name:   "bool off",
+			input:  "select %?",
+			params: []interface{}{false},
+			output: "select 0",
+			err:    "",
+		},
+		{
+			name:   "time 0",
+			input:  "select %?",
+			params: []interface{}{time.Time{}},
+			output: "select '0000-00-00'",
+			err:    "",
+		},
+		{
+			name:   "time 1",
+			input:  "select %?",
+			params: []interface{}{time.Date(2019, 1, 1, 0, 0, 0, 0, time.UTC)},
+			output: "select '2019-01-01 00:00:00'",
+			err:    "",
+		},
+		{
+			name:   "time 2",
+			input:  "select %?",
+			params: []interface{}{time2},
+			output: "select '2018-01-23 04:03:05'",
+			err:    "",
+		},
+		{
+			name:   "time 3",
+			input:  "select %?",
+			params: []interface{}{time.Unix(0, 888888888)},
+			output: "select '1970-01-01 08:00:00.888888'",
+			err:    "",
+		},
+		{
+			name:   "empty byte slice1",
+			input:  "select %?",
+			params: []interface{}{[]byte(nil)},
+			output: "select NULL",
+			err:    "",
+		},
+		{
+			name:   "empty byte slice2",
+			input:  "select %?",
+			params: []interface{}{[]byte{}},
+			output: "select _binary''",
+			err:    "",
+		},
+		{
+			name:   "byte slice",
+			input:  "select %?",
+			params: []interface{}{[]byte{2, 3}},
+			output: "select _binary'\x02\x03'",
+			err:    "",
+		},
+		{
+			name:   "string",
+			input:  "select %?",
+			params: []interface{}{"33"},
+			output: "select '33'",
+		},
+		{
+			name:   "string slice",
+			input:  "select %?",
+			params: []interface{}{[]string{"33", "44"}},
+			output: "select ('33','44')",
+		},
+		{
+			name:   "raw json",
+			input:  "select %?",
+			params: []interface{}{json.RawMessage(`{"h": "hello"}`)},
+			output: "select '{\\\"h\\\": \\\"hello\\\"}'",
+		},
+		{
+			name:   "unsupported args",
+			input:  "select %?",
+			params: []interface{}{make(chan byte)},
+			err:    "unsupported 1-th argument.*",
+		},
+		{
+			name:   "mixed arguments",
+			input:  "select %?, %?, %?",
+			params: []interface{}{"33", 44, time.Time{}},
+			output: "select '33', 44, '0000-00-00'",
+		},
+		{
+			name:   "simple injection",
+			input:  "select %?",
+			params: []interface{}{"0; drop database"},
+			output: "select '0; drop database'",
+		},
+		{
+			name:   "identifier, wrong arg",
+			input:  "use %n",
+			params: []interface{}{3},
+			err:    "expect a string identifier.*",
+		},
+		{
+			name:   "identifier",
+			input:  "use %n",
+			params: []interface{}{"table`"},
+			output: "use `table```",
+			err:    "",
+		},
+		{
+			name:   "%n missing arguments",
+			input:  "use %n",
+			params: []interface{}{},
+			err:    "missing arguments.*",
+		},
+		{
+			name:   "% escape",
+			input:  "%%?",
+			params: []interface{}{},
+			output: "%?",
+			err:    "",
+		},
+		{
+			name:   "unknown specifier",
+			input:  "%v",
+			params: []interface{}{},
+			output: "%v",
+			err:    "",
+		},
```

Comment:
How about this one?
"select * from t where val = '%%'"

Reply:
It will become `select * from t where val = '%'`, that is what we already have:
```
sql := fmt.Sprintf("UPDATE %s.%s SET variable_value='%%s' WHERE variable_name='%%s' AND variable_value=''", mysql.SystemDB, mysql.GlobalVariablesTable)
```

Reply:
And this one
"WHERE source != 'builtin'"

Reply:
Nothing, it does not have a specifier. I will add them to tests.